Implementation of Text Layout to enable multi zoom level support for win32

amartya4256 opened the pull request:
## Addressed issues
* https://github.com/eclipse-platform/eclipse.platform.swt/issues/62 
* https://github.com/eclipse-platform/eclipse.platform.swt/issues/131

## Requires
* [x] https://github.com/eclipse-platform/eclipse.platform.swt/pull/1243

**Note:** Only the last commit in this PR is to be reviewed. Previous commit(s) belong to the prerequisite PR(s)

## Description

This pull request is based on the implementations of PR #1243. This commit adds the support of scaling of text layout based on the zoom level of the monitor it is drawn on. It uses the native zoom which is provided by the GCData if available otherwise the zoom level set in the font (since the right font for a zoom level is computed already at the control level, so we always have the consistent zoom information).

amartya4256 commented:
@HeikoKlare @fedejeanne @akoch-yatta Please have a look.

akoch-yatta commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -58,14 +58,17 @@ public final class TextLayout extends Resource {
 	private MetricsAdapter metricsAdapter = new MetricsAdapter();
 
+
+	StyleItem[] allRuns;
```
Can you move that to the old position? The less changes, the better

akoch-yatta commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -355,39 +358,137 @@ void checkLayout () {
+	if (gc != null) {
+		return gc.data.nativeDeviceZoom;
+	} else if(font != null) {
+		return font.zoom;
```
Ah, that is why we set the font zoom to public. For what scenario did we need that?

akoch-yatta commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -3614,7 +3637,7 @@ public void setTextDirection (int textDirection) {
 public void setWidth (int width) {
 	checkLayout();
-	setWidthInPixels(width != SWT.DEFAULT ? DPIUtil.autoScaleUp(getDevice(), width) : width);
+	setWidthInPixels(width);
```
isn't a call to setWidthInPixel wrong now? Width is not in pixels anymore

akoch-yatta commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Font.java:
```diff
@@ -54,7 +54,7 @@ public final class Font extends Resource {
 	 * (Warning: This field is platform dependent)
 	 */
-	private int zoom;
+	int zoom;
```
Ah, wait, all fine. It is package protected. I somehow thought, it was public, my bad.

HeikoKlare commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -757,7 +759,29 @@ void drawInPixels (GC gc, int x, int y, int selectionStart, int selectionEnd, Co
+	drawInPixels(gc, DPIUtil.autoScaleUp(getDevice(), x, getZoom(gc)), DPIUtil.autoScaleUp(getDevice(), y, getZoom(gc)), selectionStart, selectionEnd, selectionForeground, selectionBackground, flags);
+}
+
+private int getNativeZoom(GC gc) {
+	if (gc != null) {
+		return gc.data.nativeDeviceZoom;
+	} else if(font != null) {
+		return font.zoom;
+	}
+	return DPIUtil.getNativeDeviceZoom();
+}
```
This looks like some incomprehensible catch-all logic to me. It simply tries to retrieve the zoom from whatever is available and otherwise falls back to a default value.
1. I do not understand why the default value makes any sense. Won't that be invalid in a multi-monitor scenario?
2. Wouldn't it be better to decide for every calling place which information _must_ be available and use that for retrieving the zoom, so that in case that fails we are sure what value is expected and may more easily find out why it is not present?

amartya4256 replied:
Have beena addressed and revised to a simpler method now.

HeikoKlare reviewed (commented):
These changes are quite hard to review, so I am wondering whether we can split them up to reduce the (accidental) complexity. Would it be possible to split this up into two subsequent commits/PRs like:
1. Replace the handling of `InPixel` values but still using DPIUtil's autoscale method
2. Replace the `autoScale` calls by `scale` calls passing the actual zoom

amartya4256 commented:
Fixed the issue with the icon of the shell. But I have also created a follow up issue which addresses the problem in the depth but it doesn't have to be the part of this PR because it addresses the usability of a method outside its scope.
https://github.com/orgs/vi-eclipse/projects/1/views/1?filterQuery=assignee%3Aamartya4256+47&pane=issue&itemId=67443386
FYI @akoch-yatta @fedejeanne @HeikoKlare

fedejeanne commented on tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/TextLayoutWin32Tests.java:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	@Test
+	public void getBoundPublicAPIshouldReturnTheSameValueRegardlessOfZoomLevel() {
```
rename to `testGetBound...`

amartya4256 replied:
done

fedejeanne commented on tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/TextLayoutWin32Tests.java:
```diff
@@ -0,0 +1,39 @@
+		int newZoom = DPIUtil.getNativeDeviceZoom() * scalingFactor;
+		GCData scaledData = new GCData();
+		scaledData.nativeZoom = newZoom;
+		GC scaledGc = GC.win32_new(display, unscaledData);
```
Shouldn't it be...
```suggestion
		GC scaledGc = GC.win32_new(display, scaledData);
```
?

amartya4256 replied:
Yes, corrected

fedejeanne commented:
> Fixed the issue with the icon of the shell. But I have also created a follow up issue which addresses the problem in the depth but it doesn't have to be the part of this PR because it addresses the usability of a method outside its scope. https://github.com/orgs/vi-eclipse/projects/1/views/1?filterQuery=assignee%3Aamartya4256+47&pane=issue&itemId=67443386 FYI @akoch-yatta @fedejeanne @HeikoKlare

Do you mean this one? https://github.com/eclipse-platform/eclipse.platform.swt/pull/1236#pullrequestreview-2115505786

amartya4256 commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/FontMetrics.java:
```diff
@@ -208,10 +215,34 @@ public int hashCode() {
+ *
+ * @noreference This method is not intended to be referenced by clients.
+ */
+public static FontMetrics win32_new(TEXTMETRIC handle, int nativeZoom) {
```
Yes

amartya4256 commented:
> > Fixed the issue with the icon of the shell. But I have also created a follow up issue which addresses the problem in the depth but it doesn't have to be the part of this PR because it addresses the usability of a method outside its scope. https://github.com/orgs/vi-eclipse/projects/1/views/1?filterQuery=assignee%3Aamartya4256+47&pane=issue&itemId=67443386 FYI @akoch-yatta @fedejeanne @HeikoKlare
> 
> Do you mean this one? [#1236 (review)](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1236#pullrequestreview-2115505786)

Yeah sorry I deleted my comment and moved it to the image PR

amartya4256 commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -1759,13 +1787,14 @@ public Rectangle getBounds () {
 	computeRuns(null);
 	int width = 0;
 	if (wrapWidth != -1) {
+		// TODO
```
No no it was something I added earlier for the ease of work but its not needed

amartya4256 commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -360,31 +363,40 @@ void checkLayout () {
 	SCRIPT_LOGATTR logAttr = new SCRIPT_LOGATTR();
 	SCRIPT_PROPERTIES properties = new SCRIPT_PROPERTIES();
-	int lineWidth = indent, lineStart = 0, lineCount = 1;
+	int wrapIndentInPixels = DPIUtil.autoScaleUp(wrapIndent, getZoom(gc));
```
Yes we would do that after we have merged everything atleast all the multi zoom level PRs so that its easy to refactor the calls. Right now they are spread across multiple PRs. And we just need to refactor them in the end.

fedejeanne replied:
Perfect, I agree.

amartya4256 commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -60,12 +60,15 @@ public final class TextLayout extends Resource {
 	StyleItem[][] runs;
-	int[] lineOffset, lineY, lineWidth;
+	int[] lineOffset, lineY;	// In Points
+	int[] lineWidth;	// In pixels
```
Talked to Andy about it. Renaming the fields would make the diff explode. I can create a separate PR for this change otherwise it would be difficult to review.

fedejeanne replied:
Perfect, do that 👍

fedejeanne reviewed (changes requested):
Tests are not running.

fedejeanne commented:
I tested this and I didn't find any regressions. As soon as the tests run in the suite (see https://github.com/eclipse-platform/eclipse.platform.swt/pull/1244#discussion_r1642455403) and all open comments from Heiko and Andreas have been marked as resolved, I will merge this PR.

amartya4256 commented on tests/org.eclipse.swt.tests.win32/JUnit Tests/org/eclipse/swt/graphics/TextLayoutWin32Tests.java:
```diff
@@ -0,0 +1,39 @@
+import org.junit.Before;
+import org.junit.Test;
+
+public class TextLayoutWin32Tests {
```
I moved it to the place where all the other tests are configured from other PRs.

fedejeanne replied:
Tests fail with this (and maybe other) error(s):

```java
Error:  org.eclipse.swt.graphics.TextLayoutWin32Tests.testGetBoundPublicAPIshouldReturnTheSameValueRegardlessOfZoomLevel -- Time elapsed: 0.057 s <<< ERROR!
java.lang.UnsatisfiedLinkError: 
Could not load SWT library. Reasons: 
	no swt-win32-4966r2 in java.library.path: /usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
	no swt-win32 in java.library.path: /usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
	no swt in java.library.path: /usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
	Can't load library: /home/runner/.swt/lib/linux/x86_64/libswt-win32-4966r2.so
	Can't load library: /home/runner/.swt/lib/linux/x86_64/libswt-win32.so
	Can't load library: /home/runner/.swt/lib/linux/x86_64/libswt.so

	at org.eclipse.swt.internal.Library.loadLibrary(Library.java:345)
	at org.eclipse.swt.internal.Library.loadLibrary(Library.java:254)
	at org.eclipse.swt.internal.C.<clinit>(C.java:19)
	at org.eclipse.swt.internal.win32.STARTUPINFO.<clinit>(STARTUPINFO.java:42)
	at org.eclipse.swt.widgets.Display.<clinit>(Display.java:149)
	at org.eclipse.swt.graphics.TextLayoutWin32Tests.setUp(TextLayoutWin32Tests.java:16)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
```

I re-triggered the Build and let it build the DLL to see if that helps

fedejeanne replied:
Perfect, it works now

fedejeanne commented on bundles/org.eclipse.swt/Eclipse SWT Tests/win32/org/eclipse/swt/graphics/TextLayoutWin32Tests.java:
```diff
@@ -0,0 +1,39 @@
+package org.eclipse.swt.graphics;
```
Copyright header is missing

amartya4256 replied:
its fixed now

fedejeanne reviewed (changes requested):
Copyright header is missing.

I also restarted the build to see if the failed tests are fixed (see my response to https://github.com/eclipse-platform/eclipse.platform.swt/pull/1244#discussion_r1642541370)
